File: packages/guards/src/disclaimer-require.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface DisclaimerRequireOptions {
  action: 'block' | 'warn';
  disclaimers?: string[];
  requireAny?: boolean;
}

const DEFAULT_DISCLAIMERS = [
  'not financial advice',
  'not medical advice',
  'not legal advice',
  'consult a professional',
  'consult a doctor',
  'consult a lawyer',
  'consult an expert',
  'for informational purposes',
  'disclaimer',
  'this is not a substitute',
  'seek professional',
  'I am an AI',
  'as an AI',
  'AI assistant',
  'not a licensed',
];

export function disclaimerRequire(options: DisclaimerRequireOptions): Guard {
  const disclaimers = options.disclaimers ?? DEFAULT_DISCLAIMERS;
  const requireAny = options.requireAny ?? true;

  return {
    name: 'disclaimer-require',
    version: '0.1.0',
    description: 'Require disclaimers in sensitive topic responses',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const found: string[] = [];

      for (const d of disclaimers) {
        if (lower.includes(d.toLowerCase())) found.push(d);
      }

      const hasDisclaimer = requireAny
        ? found.length > 0
        : found.length === disclaimers.length;

      const triggered = !hasDisclaimer;

      return {
        guardName: 'disclaimer-require',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered
          ? 'Response lacks required disclaimer for sensitive content'
          : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: {
          disclaimersFound: found,
          required: requireAny ? 'any' : 'all',
          reason: triggered
            ? 'Responses about medical, legal, or financial topics should include appropriate disclaimers'
            : undefined,
        },
      };
    },
  };
}

File: python/open_guardrail/guards/disclaimer_require.py
"""Disclaimer enforcement guard."""

import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_DISCLAIMERS = [
    "not medical advice",
    "not legal advice",
    "not financial advice",
    "consult a professional",
    "consult a doctor",
    "consult a lawyer",
    "consult an expert",
    "i am an ai",
    "as an ai",
    "ai assistant",
    "not a substitute for professional",
    "for informational purposes only",
]


class _DisclaimerRequire:
    def __init__(
        self, *, action: str = "warn", disclaimers: Optional[List[str]] = None, require_any: bool = True,
    ) -> None:
        self.name = "disclaimer-require"
        self.action = action
        self._disclaimers = [d.lower() for d in (disclaimers or _DEFAULT_DISCLAIMERS)]
        self.require_any = require_any

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        found = [d for d in self._disclaimers if d in lower]
        if self.require_any:
            triggered = len(found) == 0
        else:
            triggered = len(found) < len(self._disclaimers)
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="disclaimer-require",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="Required disclaimer not found" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"found": found, "required_count": 1 if self.require_any else len(self._disclaimers)} if triggered else None,
        )


def disclaimer_require(
    *, action: str = "warn", disclaimers: Optional[List[str]] = None, require_any: bool = True,
) -> _DisclaimerRequire:
    return _DisclaimerRequire(action=action, disclaimers=disclaimers, require_any=require_any)
